src/core/lib/iomgr/combiner.cc

```diff
@@ -228,28 +211,16 @@ bool grpc_combiner_continue_exec_ctx() {
                               grpc_core::ExecCtx::Get()->IsReadyToFinish(),
                               lock->time_to_execute_final_list));
 
-  if (contended && grpc_core::ExecCtx::Get()->IsReadyToFinish() &&
-      grpc_executor_is_threaded()) {
-    GPR_TIMER_MARK("offload_from_finished_exec_ctx", 0);
-    // this execution context wants to move on: schedule remaining work to be
-    // picked up on the executor
-    queue_offload(lock);
-    return true;
-  }
-
   if (!lock->time_to_execute_final_list ||
       // peek to see if something new has shown up, and execute that with
       // priority
       (gpr_atm_acq_load(&lock->state) >> 1) > 1) {
     gpr_mpscq_node* n = gpr_mpscq_pop(&lock->queue);
     GRPC_COMBINER_TRACE(
         gpr_log(GPR_INFO, "C:%p maybe_finish_one n=%p", lock, n));
-    if (n == nullptr) {
-      // queue is in an inconsistent state: use this as a cue that we should
-      // go off and do something else for a while (and come back later)
-      GPR_TIMER_MARK("delay_busy", 0);
-      queue_offload(lock);
-      return true;
+    while (n == nullptr) {
+      // Spinlock
```

Comment:
Could you elaborate a bit more here ? 
if I remember correctly gpr_mpscq_pop() returns `nullptr` even though its not empty (due to transient inconsistent states)  - and looks like that is what you are assuming here. That the state is transient and will soon go away.
I would like to see some comment explaining why is this a safe assumption (i.e could the queue ever be actually empty here?)